File: anders_exercises/ex01/ex01.py
# Python 3.3.3 and 2.7.6
# python helloworld_python.py

from threading import Thread

Num = 0

def incNumFunc():
    global Num
    for i in range(1000000):
    	Num = Num + 1

def decNumFunc():
	global Num
	for i in range(1000000):
		Num = Num -1

def main():
    incNumThread = Thread(target = incNumFunc, args = (),)
    incNumThread.start()

    decNumThread = Thread(target = decNumFunc, args = (),)
    decNumThread.start()
    
    incNumThread.join()
    decNumThread.join()

    print("Hello from main!")
    print("Num:" + str(Num))


main()

File: anders_exercises/ex01/ex01.c
// gcc 4.7.2 +
// gcc -std=gnu99 -Wall -g -o helloworld_c helloworld_c.c -lpthread

#include <pthread.h>
#include <stdio.h>

int Num = 0;

// Note the return type: void*
void* someThreadFunction(){
    printf("Hello from a thread!\n");
    return NULL;

}

void* decNumFunc(){
	for(int j = 0; j < 1000000; j++)
		Num++;
	return NULL;
}

void* incNumFunc(){
	for(int j = 0; j < 1000000; j++)
		Num--;
	return NULL;
	
}


int main(){
    pthread_t someThread;
    // creates a thread (scheduled by the OS)
    pthread_create(&someThread, NULL, someThreadFunction, NULL);

    pthread_t decNumThread_id;
    pthread_create(&decNumThread_id, NULL, decNumFunc, NULL);

    pthread_t incNumThread_id;
    pthread_create(&incNumThread_id, NULL, incNumFunc, NULL);


	// suspends main thread until threads in argument is completed    
    pthread_join(someThread, NULL);
    pthread_join(decNumThread_id, NULL);
    pthread_join(incNumThread_id, NULL);

    printf("Hello from main!\n");
	printf("Num: %i \n", Num);    
    return 0;
    
}
